Fix view creation on macOS (case-insensitive file system)

paulgessinger opened the pull request:
I've noticed that Spack does not correctly handle the fact that on macOS, filesystems are (usually) case-insensitive (but case-preserving).
During view creation, Spack walks the source directories of all specs and collects and deduplicates folders and files to be combined into a view. However, on macOS this can currently fail, if a folder shows up multiple times with different cases. Spack will then issue an unconditional `os.mkdir` which fails with an `OSErrror.`

I've encountered this error with the combination of the `dd4hep` and the `root` packages.

- `dd4hep` will arrange for a folder `include/ROOT` to be installed
- `root` will arrange for a folder `include/root` to be installed.

The `os.mkdir` for the second directory will fail.

I see two solutions to this:

1. Make the `os.mkdir` call conditional on whether the target folder already exists. On non-macOS platforms, in the above scenario, that would cause both `include/ROOT` and `include/root` to be created.
2. Teach the `SourceMergeVisitor` visitor to (optionally) *normalize* the paths that it checks to see if a directory (or file for. that matter) already exists.

This PR implements option 2.

The logic in this PR originally used

```python
sys.platform == "darwin"
```

to determine whether to do this or not. However, this is not correct. The case-sensitivity is a filesystem property, which is on by default in APFS, but can be turned off, for example in a separate logical volume, which people have been known to do.

This was changed to empirically check if the root of the view is on a case-insensitive filesystem, by writing to a lower case file and then seeing if it exists when asking for the upper case name.

---

Note: I think it's out-of-scope for this PR whether or not these two folders should be installed by their respective packages and only differ by case. If I build these packages using plain CMake, I will end up with a merged folder on macOS, where the actual case of the directory depends on the order they get installed. I believe Spack should reproduce this.

paulgessinger committed:
Fix view creation on macOS (case-insensitive file system)

paulgessinger committed:
Fix import ordering in `link_tree.py`

paulgessinger commented on lib/spack/llnl/util/link_tree.py:
```diff
@@ -71,10 +72,54 @@ def __init__(self, ignore: Optional[Callable[[str], bool]] = None):
+
+    def _in_directories(self, proj_rel_path: str) -> bool:
+        """
+        Check if a path is already in the direcrtory list
```
```suggestion
        Check if a path is already in the directory list
```

haampie commented:
Thanks for this PR! I will have a look on Monday. Definitely something we should get right.

Note that https://docs.python.org/3/library/os.path.html#os.path.normcase exists as a builtin

paulgessinger commented:
@haampie My understanding is that `os.path.normcase` returns the argument unmodified on Linux and Darwin. I'm guessing that's because the case sensitivity is a filesystem property and not actually a platform property.

paulgessinger commented:
@haampie Any thoughts on how I detect if the normalization is needed?

paulgessinger commented:
@haampie any thoughts on if we can move forward with this? I'm happy to discuss.

stephenswat commented:
> @haampie Any thoughts on how I detect if the normalization is needed?

I think the current solution makes a lot of sense, but if we want to be more robust, we might edit this line:

https://github.com/spack/spack/blob/2e71bc640c1ccc273b7bc71f5791316ec1832157/lib/spack/spack/filesystem_view.py#L715

And determine there whether the filesystem is case-sensitive or not; we have access to `self._root` there, so we can create a sentinel file in there, or we can even check whether `self._root` and `self._root.upper()` both exist (with the caveat that we then have to check that `self._root` has any letters in it). We could then pass that boolean down into the visitor?

paulgessinger commented:
@stephenswat yes that could work, doing it in the root directory might be enough. I was thinking about the possibility of mount points inside the view but ultimately that should not really happen.

paulgessinger committed:
Determine case-sensitivity empirically in the root dir

paulgessinger committed:
typo fix

paulgessinger commented:
@haampie what do you think of this approach?

haampie commented:
@stephenswat's suggestion is what should be done yeah, will review today.

haampie commented on lib/spack/spack/filesystem_view.py:
```diff
@@ -712,7 +712,13 @@ def add_specs(self, *specs: spack.spec.Spec) -> None:
+        # Determine if the root is on a case-insensitive filesystem
+        sentinel_name = ".sentinel"
+        with open(os.path.join(self._root, sentinel_name), "w") as f:
+            f.write("sentinel")
```
```suggestion
            pass
```

haampie replied:
The current implementation of views is to always create a fresh root, but people have requested to mutate existing dirs. So, better to not assume `.sentinel` is a file you can write to. Instead use `tempfile.mkstemp(dir=..., prefix=".sentinel")`.

Also no need to `write`, you can do `pass`.

Finally unlink the file, it should not be in the view.

paulgessinger committed:
WIP

paulgessinger committed:
revamp SourceMergeVisitor and DestinationMergeVisitor

paulgessinger committed:
add tests for source and destination merge visitors

paulgessinger committed:
call into visit_file from DestinationMergeVisitor.before_visit_symlinked_dir

paulgessinger commented:
@haampie Updated now

- I changed again how `SourceMergeVisitor` does the normalization. In the process I pulled up the branching one level like you suggested. Unfortunately, I need to keep a mapping of normalized to original projected relative path, which I implemented by the getter methods returning a tuple of three strings instead of two. This is a slight overhead for the non-normalizing path, but probably low
- I moved out the logic to determine the filesystem case sensitivity into a small utility function
  - I added a test specifically for this detection, where I mock `os.path.exists` to test both scenarios independently on host filesystem
- Updated `DestinationMergeVisitor` to call into `self.src` for all the directory and file handling. I don't need to configure this visitor manually, since it reuses the other visitor
- Added unit tests for the two visitors and combinations of files and dicts. This uses the visitor method directly, without creating any filesystem trees, because this allows it to be agnostic to the host file system properties, i.e. we can test both types of filesystem on either of the filesystems.

Let me know what you think!

paulgessinger committed:
remove unused import

paulgessinger committed:
typo fix

paulgessinger commented on lib/spack/llnl/util/link_tree.py:
```diff
@@ -253,38 +338,24 @@ def before_visit_symlinked_dir(self, root: str, rel_path: str, depth: int) -> bo
-                    rel_path, os.path.join(src_a_root, src_a_relpath), os.path.join(root, rel_path)
-                )
-            )
+        self.visit_file(root, rel_path, depth)
```
I believe this is the same logic in the two files, so I replaced it with a call to `visit_file`. Let me know if you see any issues with this.

paulgessinger committed:
whitespace

paulgessinger committed:
cleanup useless mocking context

paulgessinger committed:
fix import order

paulgessinger committed:
drop use of splitroot (python 3.12)

paulgessinger committed:
trial and error windows fix

paulgessinger commented:
I don't have a windows machine at hand so can only debug the windows test failure against the CI.

haampie commented:
cool, will have a look soon. thanks for this big effort

paulgessinger commented:
@haampie did you happen to have a look yet?

paulgessinger commented:
@haampie (bump) any news?

haampie commented on lib/spack/spack/filesystem_view.py:
```diff
@@ -884,3 +888,10 @@ def get_dependencies(specs):
+
+
+def is_folder_on_case_insensitive_filesystem(path: str) -> bool:
+    with tempfile.NamedTemporaryFile(dir=path, prefix=".sentinel") as sentinel:
+        return os.path.exists(
+            os.path.join(path, os.path.join(path, os.path.basename(sentinel.name).upper()))
+        )
```
```suggestion
    with tempfile.NamedTemporaryFile(dir=path, prefix=".sentinel") as f:
        return os.path.exists(os.path.join(path, os.path.basename(f.name).upper()))
```

one os.path.join too many

paulgessinger replied:
Fixed.

haampie commented on lib/spack/spack/test/llnl/util/link_tree.py:
```diff
@@ -396,3 +397,257 @@ def test_source_merge_visitor_does_deals_with_dangling_symlinks(tmp_path: pathli
+    return True
+
+
+def test_is_folder_on_case_sensitive_filesystem(tmp_path: pathlib.Path, monkeypatch):
```
Don't mock what you don't own. IMO it's sufficient to unit test `SourceMergeVisitor(normalize_paths=True/False)`, so you can just drop this.

paulgessinger replied:
Hm, but this function is used to determine when the visitor should enter this mode. Should I just not test that functionality?

haampie replied:
Yeah, I would just drop it. The implementation is trivial and the test is not, and the test does not help with future refactors because it relies on the implementation.

If you insist, avoid mocking by injecting the dependency as an arg: `foo(exists=os.path.exists)`. But imo the test is too complex.

paulgessinger replied:
Removed.

haampie commented on lib/spack/spack/test/llnl/util/link_tree.py:
```diff
@@ -396,3 +397,257 @@ def test_source_merge_visitor_does_deals_with_dangling_symlinks(tmp_path: pathli
+    # file conflict with os.path.samefile reporting it's the same file
+    a = SourceMergeVisitor(normalize_paths=True)
+    a.visit_file(src, "file", 0)
+    with monkeypatch.context() as m:
```
I would just create the actual files then and drop the monkeypatching. `SourceMergeVisitor` is not symbolic/abstract. You already partially create stuff in `tmp_path` anyways, and `os.path.samefile` is about inodes, not names.

paulgessinger replied:
Right but I need to test this code path in the visitor no?

The problem with using the actual file system is that the tests then depend on the file system that they run on, whereas with the direct calls to the visit function I can test this independently. 

I don't I can prepare two files with different capitalization but same inode without being on a case insensitive file system, that's why I'm using the money patch.

haampie replied:
use two directories ;)

```
src_1 = str(tmp_path / "a")
src_2 = str(tmp_path / "b")
```

paulgessinger replied:
But the filesystem access will not be case insensitive in that case will it? That means the tests can only run on macOS, does it not?

haampie replied:
I think it should be OK, cause `os.path.samefile` is about inodes, not filenames. That's the only file system operation it's doing, right?

paulgessinger replied:
I think I now understood what you were suggesting. I implemented the test without the mock now, and it works at least on macOS (but you're right it should also work on Linux I think).

haampie commented:
Looking good, I like that `DestinationMergeVisitor` doesn't need the bool

paulgessinger committed:
remove unwanted test

paulgessinger committed:
do not mock samefile, use filesystem

paulgessinger committed:
fix duplicate .join

paulgessinger commented:
@haampie I made the requested changes, can you have another look?

paulgessinger committed:
fix lint

haampie commented on lib/spack/spack/test/llnl/util/link_tree.py:
```diff
@@ -396,3 +396,230 @@ def test_source_merge_visitor_does_deals_with_dangling_symlinks(tmp_path: pathli
+
+    dest_visitor = DestinationMergeVisitor(src_visitor)
+    dest_visitor.visit_file(dest, "FILE", 0)
+    # not a conflict, since normalization is off
```
```suggestion
```

paulgessinger committed:
Update lib/spack/spack/test/llnl/util/link_tree.py

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

haampie commented:
OK, I'll save that for a follow up.
